crates/bevy_ui/src/keyboard_navigation.rs

```diff
@@ -0,0 +1,173 @@
+use crate::{FocusPolicy, Interaction, UiStack};
+use bevy_a11y::Focus;
+use bevy_ecs::prelude::Component;
+use bevy_ecs::system::{Local, Query, Res, ResMut};
+use bevy_ecs::{change_detection::DetectChangesMut, entity::Entity, query::WorldQuery};
+use bevy_input::{prelude::KeyCode, Input};
+use bevy_reflect::{FromReflect, Reflect};
+use bevy_render::view::ComputedVisibility;
+
+/// A component that represents if a UI element is focused.
+#[derive(Reflect, FromReflect, Component, Copy, Clone, Debug, Eq, PartialEq)]
+pub enum FocusedState {
+    /// Nothing has happened
+    None,
+    /// Entity is focused
+    Focus {
+        /// Focus has been reached through keyboard navigation and so a focus style should be displayed.
+        /// This is similar to the `:focus-visible` pseudo-class in css.
+        focus_visible: bool,
+    },
+}
+
+/// Should the [`keyboard_navigation_system`] run?
+pub(crate) fn tab_pressed(keyboard_input: Res<Input<KeyCode>>) -> bool {
+    keyboard_input.just_pressed(KeyCode::Tab)
+}
+
+/// Main query for [`keyboard_system`]
+#[derive(WorldQuery)]
+#[world_query(mutable)]
+pub(crate) struct KeyboardQuery {
+    interaction: Option<&'static mut Interaction>,
+    focus_policy: Option<&'static FocusPolicy>,
+    computed_visibility: Option<&'static ComputedVisibility>,
+}
+
+/// The system updates the [`Focused`] resource when the user uses keyboard navigation with <kbd>Tab</kbd> or <kbd>Shift</kbd> + <kbd>Tab</kbd>.
+///
+/// Entities can be focused [`ComputedVisibility`] is visible and [`FocusPolicy`] is block.
+pub(crate) fn keyboard_navigation_system(
+    mut focus: ResMut<Focus>,
+    mut node_query: Query<KeyboardQuery>,
```

Comment:
It's possible to have more than a single query at a time.

```rust
    mut interactions Query<&mut Interaction>,
    focusables: Query<(&FocusPolicy, &ComputedVisibility), With<Interaction>>,
```

This will avoid most of the `if Some` in this system. And I do believe it will semantically be identical.